### PR TITLE
Fix early stopping with training step's return dict

### LINKED ISSUE #3199: Early Stopping + result dictionary + no validation not working.
## 🐛 Bug

The case where the user does not use validation __and__ returns a dictionary (instead of a TrainResult) during training does not work in combination with early stopping.
The test case which should check this is here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/bd35c869ee0b856fc30229ef36957b058fe97461/tests/callbacks/test_early_stopping.py#L136-L159

The check in the last line is wrong. It should actually compare:
```python3
    assert trainer.current_epoch < trainer.max_epochs - 1
```

### To Reproduce

1. Fix the test case
2. Run tests.

#### Code sample

I guess using the test case is simpler and easier.

### Expected behavior

That is an interesting question indeed. Possibilities are:

1. Test case should pass with correct comparison
2. The [docs](https://github.com/PyTorchLightning/pytorch-lightning/blob/bd35c869ee0b856fc30229ef36957b058fe97461/docs/source/results.rst#checkpointearly_stop) and  @williamFalcon in https://github.com/PyTorchLightning/pytorch-lightning/pull/3193#discussion_r477245884 suggest that only 'loss' should work.

So before fixing this issue, it should be settled what the expected behavior is.
If you tell me, I'm happy to help.
I could also include it in the pull request where I already tried to bring the docs in line with the test cases.

### Environment

* CUDA:
	- GPU:
		- GeForce GTX 1080 Ti
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0
	- pytorch-lightning: 0.9.1dev
	- tensorboard:       2.2.0
	- tqdm:              4.48.2
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.3
	- version:           #113~16.04.1-Ubuntu SMP Fri Jul 10 04:37:08 UTC 2020

### PR DESCRIPTION
**Pull request by Lucas-Steinmann**
## What does this PR do?

Fixes #3199 

In the first commit, I fixed the test, which should check that early stopping works if no val step is executed and a return dict is used in ``training_step``. Subsequently, the test fails since the bug in #3199 is still present.

The second commit, fixes this bug, by checking the ``self.monitor`` key and not only the ``on_early_stop`` key. The aforementioned test passes.

This unveiled another bug in the decision whether early stopping is already done based on validation results.
The earlier code, checked for one key specifically. But since any key could be in validation, it is not possible to derive from the key, whether early stopping is done based on validation results. 
As a result other tests in ``test_early_stopping`` fail.

Therefore, in the third commit, I used a instance variable as flag to indicate whether early stopping is done based on validation results. The flag is set in ``on_validation_epoch_end`` which is executed before ``on_train_epoch_end``.
Lastly, I added a test which asserts that any key could be used during validation (this would have failed, if I just extended the list of keys to check besides ``val_on_early_stop``).

4th commit was necessary, since my check in the 3rd commit always disabled training early stopping if ``validation_step`` was implemented. But it should only do so if early stopping is done based on val step.

5th commit improves docs. Early stopping does not only work with ``loss`` but with any key word.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by teddykoker** [comment]
Great work! LGTM 👍

**Review by teddykoker — APPROVED** [review]

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -86,6 +86,9 @@ def __init__(self, monitor: str = 'val_loss', min_delta: float = 0.0, patience:
         self.stopped_epoch = 0
         self.mode = mode
         self.warned_result_obj = False
+        # Indicates, if eval results are used as basis for early stopping
```
Comment (nateraw) [inline pytorch_lightning/callbacks/early_stopping.py:89]: thx for the comment ❤️

**Review by nateraw — APPROVED** [review]
LGTM

**Review by nateraw — CHANGES_REQUESTED** [review]
Whoops, didn't notice a couple things...

1. gotta get code style passing
2. double check drone to make sure this works on multiple gpus. seems to failing on results. not sure if this is related to this PR or not

**Review by nateraw — APPROVED** [review]
nevermind, seems issues are unrelated. lets get this merged

**Review by SkafteNicki — APPROVED** [review]
LGTM

File: tests/callbacks/test_early_stopping.py [hunk 2]
```diff
@@ -168,3 +168,27 @@ def validation_epoch_end(self, outputs):
...
+    model = CurrentModel()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        early_stop_callback=EarlyStopping(monitor='jiraffe'),
+        overfit_batches=0.20,
+        max_epochs=20,
+    )
+    trainer.fit(model)
+    assert trainer.current_epoch == 5, 'early_stopping failed'
```
Comment (Borda) [inline tests/callbacks/test_early_stopping.py:194]:
```suggestion
    assert trainer.current_epoch >= 5, 'early_stopping failed'
```

File: tests/callbacks/test_early_stopping.py [hunk 3]
```diff
@@ -168,3 +168,27 @@ def validation_epoch_end(self, outputs):
     )
     trainer.fit(model)
     assert trainer.current_epoch == 5, 'early_stopping failed'
+
+
+def test_early_stopping_functionality_arbitrary_key(tmpdir):
+    """
+    Tests whether early stopping works with a
+    custom key and dictionary results on val step.
+    """
```
Comment (Borda) [inline tests/callbacks/test_early_stopping.py:177]:
```suggestion
    """Tests whether early stopping works with a custom key and dictionary results on val step."""
```

**Review by Borda — APPROVED** [review]
